### PR TITLE
map, filter & reduce collection helpers

### PR DESCRIPTION
**Pull request by AydinHassan**
### Map, Filter & Reduce functions for Collections.

We all know it is impossible to use the `array_` functions from PHP std library with user land collections. This PR is purely syntactical sugar, as there are other ways to achieve the same results without these additions.

However, providing these methods is fairly common practise in userland collection implementations and can be immensely useful, see these popular implementations: 
- https://github.com/doctrine/collections/blob/master/lib/Doctrine/Common/Collections/Collection.php
- https://github.com/illuminate/support/blob/master/Collection.php
- https://github.com/morrisonlevi/Ardent
### Some examples:

``` php
$collection = new \Magento\Framework\Data\Collection();
$collection->addItem($item1);
$collection->addItem($item2);

$collection->map(function ($item) {
    return $item->setData('some_key', 'some_value');
});

$collection->filter(function ($item) {
    return $item->getData('price') > 0;
});

$total = $collection->reduce(function ($carry, $item) {
    return $carry + $item->getData('price');
}, 0);
```

It is possible to achieve this already, but the process is rather cumbersome and requires much more code and more variables. This more functional approach is clean and concise. 

I see this as really useful for just manipulating collections without having to convert them to their basic PHP array format. Code becomes more expressive and it's intent is clear through these API's.
### Caveats
1. There are quite a few methods with `filter` in there name in the various collections throughout Magento. The method `filter` may be slightly ambiguous. I am open for suggestions - however, this name is used in many programming languages and libraries for this very purpose.
2. There are already ways to get the same results.
### Future

We could add more methods in the future for a slick, clean OO API, such as `contains`, `slice`, `chunk` and so on, or we add nothing more at all.

### CODE REVIEW

File: lib/internal/Magento/Framework/Data/Collection.php [hunk 1]
```diff
@@ -527,6 +527,42 @@ public function each($objMethod, $args = [])
...
+        $this->_items = array_map($callback, $this->_items);
+    }
+
+    /**
+     * Applies the given callable to each element in the collection and removes items
+     * which do not pass the callable criteria.
+     *
+     * @param callable $callback
+     */
+    public function filter(callable  $callback)
```
Comment (adragus-inviqa) [inline lib/internal/Magento/Framework/Data/Collection.php]: /cc https://github.com/magento/magento2/pull/2026

**Comment by adragus-inviqa** [comment]
Although nice, you're adding methods to the public API of a quite common class; so it seems risky.

I would rather have an `EnhancedCollectionDecoratorSomething` class implementing an `EnhancedCollectionInterfaceContractSomething`.

PS: I had a couple of attempts adding public methods, too, but got rejected on the basis you are now.

**Comment by AydinHassan** [comment]
@adragus-inviqa I understand your concerns, however, for me, your suggestion would negate completely the benefits this brings (for me).

A new collection class would not allow me to use these methods on the majority of collections which already extend from this class. If I were to write a new collection class I may as well just bring in a well tested already built package. I would also still have to move the items, or wrap the existing collections. That seems too complex for just a bit of sugar.

Interesting, I did search but nothing came up - still, worth a shot 😄 .

**Comment by adragus-inviqa** [comment]
Maybe they'll think otherwise of your proposal, which is indeed useful.

Maybe `filterBy()`, `mapWith()` etc.? I'm not fond of short names, however established they might be.

I've been told this is quite good, as well: http://book.cakephp.org/3.0/en/core-libraries/collections.html

PS: If they ok this, I'd bring the "future" now, not wait for another version.

**Comment by Vinai** [comment]
I like your idea (quite a lot), but would like to propose a change.
Rather then modifying `$_items` of the collection in filter and map, how about returning a new collection instance with the result set, leaving the collection the method was called on unchanged.
This would be more in line with any other implementation of map/reduce I've come across so far, and it is more in line with the functional programming paradigm where map/reduce originated.

**Comment by AydinHassan** [comment]
@Vinai your suggestion is completely warranted - and if I were implementing this myself in one of my own projects, I would do it that way. I pondered implementing it here but I felt it was out of place with Magento's usual mutable process. 

If the majority agreed this was the favourable approach, then I would totally be on board too.

Maybe we could implement it like so:

``` php
/**
 * Applies the given callable to each element in the collection.
 *
 * @param callable $callback
 * @return static
 */
public function map(callable $callback)
{
    $this->load();

    $new = clone $this;
    $new->_items = array_map($callback, $this->_items);
    return $new;
}
```

``` php
/**
 * Applies the given callable to each element in the collection and removes items
 * which do not pass the callable criteria.
 *
 * @param callable $callback
 * @return static
 */
public function filter(callable  $callback)
{
    $this->load();

    $new = clone $this;
    $new->_items = array_filter($this->_items, $callback);
    return $new;
}
```

**Comment by katmoon** [comment]
@antonkril could you take a look at this PR, please?

**Comment by adragus-inviqa** [comment]
Actually, some of my more knowledgeable colleagues started crying when they heard that this implementation is changing the same collection. So let's push the immutable card; see what they say.

**Comment by AydinHassan** [comment]
@adragus-inviqa @Vinai I've updated the PR to follow the immutable approach, interestingly though: if you modify objects in the map, then the changes will be seen in the original collection as well as the newly returned collection - I think it makes it a bit counter intuitive - what are your thoughts?

**Comment by adragus-inviqa** [comment]
I think it's because it's a shallow-clone. All the cloning won't matter if `_items` contains the same references. 

And even if you iterate, cloning each object in `_items`, you'd still need to clone _their_ object properties and then watch out for double linking, which will give you an infinite loop. And the seriliaze/unserialize option is too expensive for a big object graph.

So... yeah.

**Comment by AydinHassan** [comment]
Yeah I mean it's not really a problem - it would be the same in any other collection library - I think it's just because this collection class can only contain objects it seems a bit weird. In any case you don't really use `map` to modify items, so this is a non issue - maybe I'll update the test, because I think the operation is a bit strange and not something you would really do with a map, it's more a `walk/each` operation. Something like the following is a bit more illustrative:

``` php
$prices = $products->map(function (\Magento\Framework\DataObject $obj) {
    return $obj->getData('price');
});
```

Then again - this fill the collection with items it should not contain (you can't manually add anything that does not inherit from `DataObject` ) - you mentioned this before. So I'm not really sure what we should do here - maybe `map` doesn't work for this scenario.

**Comment by ghost** [comment]
internal ticket created: MAGETWO-56734

**Comment by vkublytskyi** [comment]
@AydinHassan, thank you for contribution. Collections now overloaded with logic so we have to add new methods to it only with strong reasons.
As Collection has method `getItems` array functions may be used in almost all cases.

**Pull request closed without merging**